### ISSUE TITLE
How to run the sgbm node? 

### ISSUE THREAD

**Issue by ashnarayan13** [ShreyasSkandan/ros_tegra_stereo#1]
Hi, I have cloned the repo and could build the package. But I am not sure how to run the sgbm node. What parameters should I pass to the node? 

Thanks!

**Comment by ShreyasSkandan** [comment]
Hi, if you're trying to run my repository, you need to pass a set of camera parameters as arguments in the same form as the file **data/cam_stereo.yml**

I have a multiple baseline setup, which is why I usually pass two sets of parameter files. You can specify which baseline you'd like to use with the **baseline_opt** parameter.

The arguments are defined as follows,

`  static struct poptOption options[] = {
  { "b1_param_file",'c',POPT_ARG_STRING,&b1_param_file,0,"Destination for b1 calib file","STR" },
  { "b2_param_file",'f',POPT_ARG_STRING,&b2_param_file,0,"Destination for b2 calib file","STR" },
  { "baseline_opt",'b',POPT_ARG_INT,&baseline_opt,0,"Baseline option","NUM"},
  POPT_AUTOHELP
  { NULL, 0, 0, NULL, 0, NULL, NULL }
  };`

Let me know if you have any more questions,

- Shreyas

**Comment by ashnarayan13** [comment]
Thanks! 
I want to run it for stereo cameras. So I just replace the cam_stereo.yml with my camera calibration and run the sgbm_node? 
rosrun ros_tegra_stereo sgbm_node

**Comment by ShreyasSkandan** [comment]
Yes, but you also need to pass the appropriate parameters over command line. Please look at the poptOption options[] datastructure to see what arguments need to be passed. It generally will look like this,

rosrun ros_tegra_stereo sgbm_node -c "path/to/calib_baseline_1" -f "path/to/calib_baseline_2" -b "[0/1]"

**Comment by ashnarayan13** [comment]
I get a black image as the disparity. I checked if the images were loaded correctly by ros and they were. I am not sure where the error is. 
I just removed the image rectification part as I already provide rectified images and hence only need stereo done. (replaced reading a single image with reading 2 images using message filters)
Any suggestions as to where the error might be?

**Comment by ashnarayan13** [comment]
This is my updated callback method
`cv_bridge::CvImagePtr cv_left;
  try
    {
      cv_left = cv_bridge::toCvCopy(left);
    }
    catch (cv_bridge::Exception& e)
    {
      ROS_ERROR("cv_bridge exception: %s", e.what());
      return;
    }

  cv_bridge::CvImagePtr cv_right;
  try
      {
        cv_right = cv_bridge::toCvCopy(right);
      }
      catch (cv_bridge::Exception& e)
      {
        ROS_ERROR("cv_bridge exception: %s", e.what());
        return;
      }
	img1 = cv_left->image;

    left_rect = nvx_cv::createVXImageFromCVMat(context,cv_left->image);
    NVXIO_CHECK_REFERENCE(left_rect);
    right_rect = nvx_cv::createVXImageFromCVMat(context,cv_right->image);
    NVXIO_CHECK_REFERENCE(right_rect);
	ROS_INFO("%d %d", img1.cols, img1.rows);
    disparity = vxCreateImage(context, IMG_WIDTH, IMG_HEIGHT, VX_DF_IMAGE_U8);
    NVXIO_CHECK_REFERENCE(disparity);
    std::unique_ptr<StereoMatching> stereo(StereoMatching::createStereoMatching(context,
					   params,implementationType,left_rect, right_rect, disparity));
    stereo->run(); 
    nvx_cv::VXImageToCVMatMapper map(disparity,plane_index,&rect,VX_READ_ONLY,VX_MEMORY_TYPE_HOST);
    cv::Mat disp = map.getMat();
	imwrite("disp.png",disp);
	ROS_INFO("here?");
    //cv::imshow("disparity",disp);
	//imwrite("left.png",cv_left->image);
	//imwrite("right.png", cv_right->image);


    cv_bridge::CvImage out_disp;
    out_disp.header = cv_left->header;
    out_disp.encoding = sensor_msgs::image_encodings::MONO8;
    out_disp.image = disp;
    pub.publish(out_disp.toImageMsg());`

My updated main() is as follows
`nvxio::Application &app = nvxio::Application::get();
  app.setDescription("Stereo Matching for Disparity Estimation");

  app.setDescription("Stereo Matching for Disparity Estimation");
  std::string configFile = "/home/ubuntu/catkin_ws/src/ros_tegra_stereo/data/stereo_matching_params.ini"; // change

  implementationType = StereoMatching::HIGH_LEVEL_API;
  std::string error;
  if (!read(configFile, params, error)) {
      std::cerr << error;
      return 0;
  }
  std::cout << "SGBM Parameters loaded\n" ;

  vxDirective(context, VX_DIRECTIVE_ENABLE_PERFORMANCE);
  vxRegisterLogCallback(context, &ovxio::stdoutLogCallback, vx_false_e);

  left_rect = vxCreateImage(context, IMG_WIDTH, IMG_HEIGHT, VX_DF_IMAGE_RGBX);
  NVXIO_CHECK_REFERENCE(left_rect);
  right_rect = vxCreateImage(context, IMG_WIDTH, IMG_HEIGHT, VX_DF_IMAGE_RGBX);
  NVXIO_CHECK_REFERENCE(right_rect);
  disparity = vxCreateImage(context, IMG_WIDTH, IMG_HEIGHT, VX_DF_IMAGE_U8);
  NVXIO_CHECK_REFERENCE(disparity);

  image_transport::ImageTransport it(nh);
//message_filters for left/right images
  message_filters::Subscriber<Image> left_sub(nh, "/camera/left/image_rect", 1);
  message_filters::Subscriber<Image> right_sub(nh, "/camera/right/image_rect", 1);

  //time syncronizer to publish 2 images in the same callback function
  typedef sync_policies::ApproximateTime<Image, Image> MySyncPolicy;
  Synchronizer<MySyncPolicy> sync(MySyncPolicy(1), left_sub, right_sub);
  //TimeSynchronizer<Image, Image> sync(left_sub, right_sub, 50);

  //call calback each time a new message arrives
  sync.registerCallback(boost::bind(&imageCallback, _1, _2));
  pub = it.advertise("/disparityimage",1); // change`

**Comment by ShreyasSkandan** [comment]
That's strange. So the disparity image returned to you is all 0s?

**Comment by ashnarayan13** [comment]
yes. It does computation and takes time and publishes black images all 0. I removed the popt option as I provide rectified images already. So I am not sure where the issue is occurring. I checked if the code enters the vision works compute and it enters the computing stereo image part but still provides me these black images. 

Any suggestions?

**Comment by ashnarayan13** [comment]
is there a ROSBAG which you used to test that could be added so that I can test what is wrong in my implementation?

**Comment by ShreyasSkandan** [comment]
Hi @ashnarayan13, sorry for not getting back to you sooner. 
I have updated the repository now such that it only takes in stereo images as input. This should be more suited towards your application. There is also the option of publishing as a stereo_msgs::DisparityImage. You can test it out by using image_view's disparity_view. Let me know if this helps. If not, I can put together a test rosbag for you.

Best regards,
Shreyas
